### ISSUE TITLE
Unparseable section with Databricks function when parameters contain functions

### ISSUE THREAD

**Issue by avinmathew** [sqlfluff/sqlfluff#7240]
### Search before asking

- [x] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

I receive a "Found unparsable section" when adding a function as a default value to the function parameter.

### Expected Behaviour

Parsing should not fail when there is a function in the Databricks function parameters, given that the code executes successfully Databricks.

### Observed Behaviour

I received the error:
```
L:   1 | P:   1 |  PRS | Line 1, Position 1: Found unparsable section: 'CREATE
                       | FUNCTION example_function (\n    p...'
```

### How to reproduce

This is a minimal SQL will reproduce the problem with either linting or fixing:
```sql
CREATE FUNCTION example_function (
    param1 date DEFAULT current_date()
)
RETURNS TABLE
RETURN
SELECT param1 AS col;
```

Removing the DEFAULT works without error:
```sql
CREATE FUNCTION example_function (
    param1 date
)
RETURNS TABLE
RETURN
SELECT param1 AS col;
```

Having a DEFAULT with a date strings also works without error:
```sql
CREATE FUNCTION example_function (
    param1 date DEFAULT DATE 'today'
)
RETURNS TABLE
RETURN
SELECT param1 AS col;
```

### Dialect

Databricks

### Version

sqlfluff, version 3.5.0
Python 3.11.9

### Configuration

[sqlfluff]
dialect = databricks

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [x] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)
